torchaudio_contrib/functional.py

```diff
@@ -1,106 +1,133 @@
-        out_shape = [batch, channel, n_fft//2+1, -1, 2]
-    elif sig.dim() == 2:
-        channel, time = sig.size()
-        out_shape = [channel, n_fft//2+1, -1, 2]
+
+    # (!) Only 3D, 4D, 5D padding with non-constant
+    # padding are supported for now.
+    if pad > 0:
+        x = F.pad(x, (pad, pad), pad_mode)
+
+    if x.dim() == 3:
+        batch, channel, time = x.size()
+        out_shape = [batch, channel, n_fft // 2 + 1, -1, 2]
+    elif x.dim() == 2:
+        channel, time = x.size()
+        out_shape = [channel, n_fft // 2 + 1, -1, 2]
     else:
         raise ValueError('Input tensor dim() must be either 2 or 3.')
 
-    sig = sig.reshape(-1, time)
-    stft = torch.stft(sig, n_fft, hop, window=window, **kwargs)
-    stft = stft.reshape(out_shape)
+    x = x.reshape(-1, time)
+
+    stft_out = torch.stft(x, n_fft, hop_length, window=window,
+                          win_length=window.size(0), **kwargs)
+    stft_out = stft_out.reshape(out_shape)
+
+    return stft_out
+
 
-    return stft
+def stft(x, n_fft=2048, hop_length=None, len_win=None,
+         window=None, pad=0, pad_mode="reflect", **kwargs):
+    """
+    Wraps _stft setting default values and correct window device.
+    See torchaudio_contrib.STFT for more details.
+    """
+    n_fft, hop_length, window = stft_defaults(
+        n_fft, hop_length, len_win, window)
+    return _stft(x, n_fft=n_fft, hop_length=hop_length,
+                 window=window.to(x.device), pad=pad,
+                 pad_mode=pad_mode, **kwargs)
 
 
-def spectrogram(sig, n_fft=2048, hop=None, window=None, power=1.0, **kwargs):
+def complex_norm(x, power=1.0):
+    """
+    Normalize complex tensor.
     """
-    returns magnitude of spectrogram
+    return x.pow(2).sum(-1).pow(power / 2.0)
 
-    input -> (batch, channel, time) or (channel, time)
-    output -> (batch, channel, freq, hop) or (channel, freq, hop)
+
+def spectrogram(x, n_fft=2048, hop_length=None, len_win=None,
+                window=None, pad=0, pad_mode="reflect", power=1., **kwargs):
+    """
+    Compute the spectrogram of a given signal.
+    See torchaudio_contrib.Spectrogram for more details.
     """
-    stft = STFT(sig, n_fft, hop=hop, window=window, **kwargs)
-    if power is None:
-        return stft
-    return stft.pow(2).sum(-1).pow(power / 2.0)
+    return complex_norm(stft(x, n_fft, hop_length, len_win,
+                             window, pad, pad_mode, **kwargs), power=power)
 
 
 def _hertz_to_mel(f):
-    '''
+    """
     Converting frequency into mel values using HTK formula
-    '''
+    """
     return 2595. * torch.log10(torch.tensor(1.) + (f / 700.))
 
 
 def _mel_to_hertz(mel):
-    '''
+    """
     Converting mel values into frequency using HTK formula
-    '''
+    """
     return 700. * (10**(mel / 2595.) - 1.)
 
 
-def melspectrogram(sig, n_mels=128, sr=44100, f_min=0.0, f_max=None, *args, **kwargs):
+def create_mel_filter(n_mels, sr, f_min, f_max, n_stft):
```

Comment:
In the interest of reusability, I think it should take the mel/hz conversion functions as an input.

Reply:
Right... when you say reusability here is a use case like the pitch shifting augmentation from the other thread? I think that would make sense, but then how would `Melspectrogram` set the conversion functions? Are the parameters of the conversion functions something that `Melspectrogram` should be able to change dynamically?

Reply:
That's a point where I find it confusing to think both about the functional and the object-oriented interface. In the latter, I'd say we'd have a module that applies a filterbank, this holds a module that provides a filterbank, and this is a base class with a subclass for each mel flavor and for CQT that only have to override the `to_hz` and `from_hz` methods, while the base class handles storing the filter bank as a buffer and providing it in its `forward()` method. The pitch shifting augmentation would be a wrapper class that overrides the `to_hz` and `from_hz` functions to apply some scaling and computes the filter bank on-the-fly in its `forward()` method. The tricky part is where it gets the scalings from, and if it should support applying a different one per batch item.
If we want to provide different filter shapes (not only triangular filters), and we want this independently of the scale, we'd end up with either a hefty base class with several `if`s or a second class hierarchy, but maybe we shouldn't overdo it... judging from what the model zoo looks like, the spirit of PyTorch seems to be to copy and paste code rather than making it easy to reuse things ;)

Reply:
>  I'd say we'd have a module that applies a filterbank, this holds a module that provides a filterbank

Something like [this](https://github.com/keunwoochoi/torchaudio-contrib/blob/af20c5bd60badcb8ab82ee83d1eb746c8e43a0e5/torchaudio_contrib/layers.py#L127)?

Then when calling `Melspectrogram`, by default `MelFilterbank` [will be used](https://github.com/keunwoochoi/torchaudio-contrib/blob/af20c5bd60badcb8ab82ee83d1eb746c8e43a0e5/torchaudio_contrib/layers.py#L297) but can also pass in a subclass with overridden `to_hertz`/`from_hertz` per:
>   this is a base class with a subclass for each mel flavor and for CQT that only have to override the `to_hz` and `from_hz` methods